File: pys/dt.py
# thebell 거의 일치
# 인코딩 : iso-8859-1로 인코딩 후 euc-kr로 디코딩

import requests
from bs4 import BeautifulSoup

url = 'http://www.dt.co.kr/section.html?section_num=2900'
req = requests.get(url)
bs = BeautifulSoup(req.text, 'lxml')

wrapper = bs.find('div', class_='list_area')
items = wrapper.find_all('dl', class_='article_list')

for item in items[:3]:
    dt = item.find('dt')
    a_tag = dt.find('a')
    href = '' + a_tag['href']
    title = a_tag.get_text().strip().encode('iso-8859-1').decode('euc-kr')
    date = item.find('span', class_='date').get_text().encode('iso-8859-1').decode('euc-kr').replace('입력 ', '').strip()

    req2 = requests.get(href)
    bs2 = BeautifulSoup(req2.text, 'lxml')

    wrapper2 = bs2.find('div', class_='art_txt')
    contents = ''.join([it.strip() for it in wrapper2.find_all(text=True, recursive=False)]).encode('iso-8859-1').decode('euc-kr')

    print(title, href, date, contents)

File: crawlers/dt.go
package crawlers

import (
	"fmt"
	"net/http"
	"strings"

	"flashnews/models"
	"flashnews/utils"

	"github.com/PuerkitoBio/goquery"
)

const dtCommonURL = "http://www.dt.co.kr/section.html?section_num=2900"
const dtItemURL = ""

type DT struct{}

func (c DT) GetName() string {
	return "dt"
}

func (c DT) GetGroup() string {
	return "4"
}

func (c DT) GetList(number int) ([]models.NewsItem, error) {
	// Number
	var _number int
	if number > 10 || number < 1 {
		_number = 10
	} else {
		_number = number
	}
	result := make([]models.NewsItem, _number)

	// Request
	req, err := http.Get(dtCommonURL)
	if err != nil {
		return result, err
	}
	defer req.Body.Close()

	if req.StatusCode != 200 {
		return result, fmt.Errorf("[ERROR] Request Status Code : %d, %s", req.StatusCode, req.Status)
	}

	// HTML Load
	html, err := goquery.NewDocumentFromReader(req.Body)
	if err != nil {
		return result, err
	}

	// Parsing
	wrapper := html.Find("div.list_area")
	items := wrapper.Find("dl.article_list")
	items.Each(func(i int, sel *goquery.Selection) {
		if i >= _number {
			return
		}

		dtTag := sel.Find("dt")
		aTag := dtTag.Find("a")
		href, ok := aTag.Attr("href")
		if !ok {
			result[i] = models.NewsItem{}
			return
		}
		url := dtItemURL + href

		date, err := utils.ReadISO88591(sel.Find("span.date").Text())
		if err != nil {
			result[i] = models.NewsItem{}
			return
		} else {
			date = strings.Replace(date, "입력 ", "", -1)
		}

		title, err := utils.ReadISO88591(aTag.Text())
		if err != nil {
			result[i] = models.NewsItem{}
			return
		}

		result[i] = models.NewsItem{
			Title:    title,
			URL:      url,
			Contents: "",
			Datetime: date,
		}
	})

	return result, nil
}

func (c DT) GetContents(item *models.NewsItem) error {
	// Request
	req, err := http.Get(item.URL)
	if err != nil {
		return err
	}
	defer req.Body.Close()

	if req.StatusCode != 200 {
		return fmt.Errorf("[ERROR] Request Status Code : %d, %s", req.StatusCode, req.Status)
	}

	// HTML Load
	html, err := goquery.NewDocumentFromReader(req.Body)
	if err != nil {
		return err
	}

	// Parsing
	wrapper := html.Find("div.art_txt")
	contents := ""
	wrapper.Contents().Each(func(i int, sel *goquery.Selection) {
		if goquery.NodeName(sel) == "#text" {
			temp, err := utils.ReadISO88591(utils.TrimAll(sel.Text()))
			if err != nil {
				temp = ""
			}
			contents += (utils.TrimAll(temp) + " ")
		}
	})

	item.Contents = contents
	return nil
}
